plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/concurrent/LSPExecutorService.java

```diff
@@ -0,0 +1,296 @@
+ */
+package org.eclipse.n4js.ide.xtext.server.concurrent;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.log4j.Logger;
+import org.eclipse.xtext.service.OperationCanceledManager;
+import org.eclipse.xtext.util.CancelIndicator;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * Similar to an ordinary {@link ExecutorService}, but with two differences:
+ * <ol>
+ * <li>when submitting a task, a queue ID may be provided and this executor will then guarantee that
+ * <ol type="a">
+ * <li>two tasks with {@link Object#equals(Object) equal} queue ID are never running at the same time, and
+ * <li>tasks with equal queue ID will be executed in the order they were submitted.
+ * </ol>
+ * If no queue ID is provided, the usual behavior of ordinary {@link ExecutorService executor services} applies, i.e.
+ * the task might run concurrently to any other task and ordering depends on the {@link #delegate delegate executor
+ * service}.
+ * <li>each task is provided with an Xtext {@link CancelIndicator} and cancellation via a task's
+ * {@link CompletableFuture future}, as returned by the submit methods, is converted into a cancellation request via
+ * that cancel indicator. This leads to a slight change how cancellation works compared to ordinary futures, see
+ * {@link QueuedTaskFuture#cancel(boolean) #cancel(boolean)} for details.
+ * </ol>
+ *
+ * <h2>Cancellation Handling</h2>
+ *
+ * Cancellation is handled entirely through Xtext's {@link CancelIndicator}s and a task's implementation can decide how
+ * to react to that (as usual with cancel indicators). Tasks will always be invoked, even if they were marked as
+ * cancelled while waiting on the queue, so implementors of a task may choose to implement a "non-cancellable" operation
+ * at the beginning of a task before reacting to the cancel indicator.
+ */
+// FIXME remove @Singleton?
+@Singleton // ensures queue IDs are "global" within a single injector
+public class LSPExecutorService {
+
+	private static final Logger LOG = Logger.getLogger(LSPExecutorService.class);
+
+	/** The underlying executor service that is used to actually execute the tasks. */
+	@Inject
+	protected ExecutorService delegate;
+
+	/***/
+	@Inject
+	protected OperationCanceledManager operationCanceledManager;
+
+	/** Global queue of all pending tasks across all queue IDs. */
+	protected final List<QueuedTask<?>> pendingTasks = new ArrayList<>();
```

Comment:
This should probably be an ArrayDeque

Reply:
`#pollNextPending()` is iterating over the elements and removing elements from other places than just head or tail, so its more used like a list.